File: tournaments/concatenationProcess/concatenationProcess.cs
string concatenationProcess(string[] initialArray) {
    var r = new List<string>(initialArray);
    
    while (r.Count > 1) {
        int min = int.MaxValue;
        int mini = -1;
        
        for (int i = 0; i < r.Count; i++) {
            if (r.ElementAt(i).Length < min) {
                min = r.ElementAt(i).Length;
                mini = i;
            }
        }
        
        int min2 = int.MaxValue;
        int mini2 = -1;
        for (int i = r.Count - 1; i >= 0; i--) {
            if (i != mini && r.ElementAt(i).Length < min2) {
                min2 = r.ElementAt(i).Length;
                mini2 = i;
            }
        }
        
        string neu = r.ElementAt(mini) + r.ElementAt(mini2);
        
        if (mini < mini2) {
            r.RemoveAt(mini2);
            r.RemoveAt(mini);
        } else {
            r.RemoveAt(mini);
            r.RemoveAt(mini2);
        }
        
        r.Add(neu);
    }
    
    return r.First();
}

File: tournaments/concatenationProcess/concatenationProcess.cpp
std::string concatenationProcess(std::vector<std::string> initialArray) {

  while (initialArray.size() > 1) {
    int minInd1 = 0,
        minInd2 = initialArray.size() - 1;

    for (int i = 1; i < initialArray.size(); i++) {
      if (initialArray[i].size() < initialArray[minInd1].size()) {
        minInd1 = i;
      }
    }

    if (minInd2 == minInd1) {
      minInd2--;
    }

    for (int i = initialArray.size() - 2; i >= 0; i--) {
      if (initialArray[i].size() < initialArray[minInd2].size()
       && i != minInd1) {
        minInd2 = i;
      }
    }

    initialArray.push_back(initialArray[minInd1] + initialArray[minInd2]);
    initialArray.erase(initialArray.begin() + std::max(minInd1, minInd2));
    initialArray.erase(initialArray.begin() + std::min(minInd1, minInd2));
  }

  return initialArray[0];
}
